_posts/2021-12-07-g2p-case-feeding.md

```diff
@@ -0,0 +1,143 @@
+
+Imagine we use a g2p mapping to modify some spelling rules, and the same sequence of letters might get mapped differently depending on the context it is found in. In our minimal example, we want to handle "in" as follows:
+ - when the string "atin" occurs, it should be replaced by "etin", and that "in" should not be further modified;
+ - when "in" occurs before "a" or "e", it should remain as "in";
+ - otherwise, "in" should be changed to "an" (catch-all rule).
+
+With these rules, the word "intinatin" should get changed to "antinetin" because the third "in" is part of "atin", the second "in" is followed by "a" or "e", and only the first "in" is handled by our catch-all rule.
+
+Notice that the first two rules preserve "in" in their output, so we have to make sure the output of a rule cannot be reused as the output of a subsequent rule.
+
+## The obvious solution, which does not work
+
+We can almost solve this by using `prevent_feeding`, but not quite.
+
+Let's write the rules as follows:
+
+`mapping.csv`:
+```
+atin,etin,,
+in,in,,[ae]
+in,an,,
+```
+
+and use config file `config.yaml`, where we are careful to set `prevent_feeding: true`:
+```
+display name: mapping example for in with prevent feeding
+mapping: mapping.csv
+in_lang: in
+out_lang: out
+case_sensitive: false
+prevent_feeding: true
+```
+
+When you run `g2p convert --config config.yaml intinatin in out`, you will get "antanetin" as output instead of "antinetin".
+
+**What's the error?** (I know, it's subtle...) The middle "in" should have matched the second rule, since it was followed by "a" before the first rule was applied, and is still followed by "e" after the first rule was applied. That should have blocked the application of the third rule, but apparently it didn't.
+
+**Why is this happening?** We need to talk about how `prevent_feeding` is implemented to answer that question. The point of the prevent-feeding option is to make sure that the output of a rule is never matched as the input of any other rule. To accomplish that, we actually map the output of each rule temporarily to characters in a [Private Use Area](https://en.wikipedia.org/wiki/Private_Use_Areas) in the Unicode standard, and map them back to the real output once all rules have been applied. The private use area characters are intended for internal (hence "private") use within software, but should never be printed, so they were perfect to solve the prevent-feeding problem: they should never occur in input text or in the input of any rule, and so there would never unintended feeding between rules.
+
+The problem is that those private characters are also inaccessible to the `context_before` and `context_after` parts of our rules. So `prevent_feeding` not only blocks the characters from being matched as the input of other rules, it also blocks them from being matched in their contexts.
```

Comment:
I wonder if you could suggest that the user use the `--debugger` flag to watch this in action